snappy-core/src/main/scala/org/apache/spark/sql/store/JDBCSourceAsStore.scala

```diff
@@ -143,36 +144,70 @@ class JDBCSourceAsStore(override val connProperties: ConnectionProperties,
   }
 }
 
-final class CachedBatchIteratorOnRS(conn: Connection,
-    requiredColumns: Array[String],
-    ps: Statement, rs: ResultSet) extends Iterator[CachedBatch] {
+abstract class ResultSetIterator[A](conn: Connection,
+    stmt: Statement, rs: ResultSet, context: TaskContext)
+    extends Iterator[A] with Logging {
 
-  var _hasNext = moveNext()
+  protected final var hasNextValue = true
 
-  override def hasNext: Boolean = _hasNext
+  context.addTaskCompletionListener { context => close() }
+  moveNext()
 
-  override def next() = {
-    val result = getCachedBatchFromRow(requiredColumns, rs)
-    _hasNext = moveNext()
-    result
-  }
+  override final def hasNext: Boolean = hasNextValue
 
-  private def moveNext(): Boolean = {
+  protected final def moveNext(): Unit = {
     var success = false
     try {
+      // TODO: see if optimization using rs.lightWeightNext
+      // and explicit context pop in close possible (was causing trouble)
       success = rs.next()
-      success
+    } catch {
+      case NonFatal(e) => logWarning("Exception iterating resultSet", e)
     } finally {
       if (!success) {
-        rs.close()
-        ps.close()
-        conn.close()
+        close()
       }
     }
   }
 
-  private def getCachedBatchFromRow(requiredColumns: Array[String],
-      rs: ResultSet): CachedBatch = {
+  protected def getNextValue(rs: ResultSet): A
+
+  final def next(): A = {
+    val result = getNextValue(rs)
+    moveNext()
+    result
+  }
+
+  final def close() {
+    if (!hasNextValue) return
+    try {
+      // GfxdConnectionWrapper.restoreContextStack(stmt, rs)
+      // rs.lightWeightClose()
+      rs.close()
+    } catch {
+      case e: Exception => logWarning("Exception closing resultSet", e)
+    }
+    try {
+      stmt.close()
+    } catch {
+      case e: Exception => logWarning("Exception closing statement", e)
+    }
+    try {
+      conn.close()
+      logDebug("closed connection for task " + context.partitionId())
+    } catch {
+      case e: Exception => logWarning("Exception closing connection", e)
+    }
+    hasNextValue = false
+  }
+}
+
+final class CachedBatchIteratorOnRS(conn: Connection,
+    requiredColumns: Array[String],
+    stmt: Statement, rs: ResultSet, context: TaskContext)
+    extends ResultSetIterator[CachedBatch](conn, stmt, rs, context) {
+
+  protected override def getNextValue(rs: ResultSet): CachedBatch = {
     // it will be having the information of the columns to fetch
     val numCols = requiredColumns.length
```

Comment:
Thanks @rishitesh. I will go ahead with merge of this because zoomdata is waiting for fix for this and I have subsequent fixes lined up for other reported issues. Can you update your PR after this merge?